### PR TITLE
Keep track of desired nodes cluster membership

### PR DESCRIPTION
**Pull request by fcofdez**
This commit adds tracking for desired nodes cluster membership.

When desired nodes are updated they are matched against the current
cluster members. Additionally when a node joins the cluster the
desired nodes cluster membership is updated.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java:143]: I think we should take into account previous desired nodes marked as members. I'll fix it.

Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java]: Tackled in https://github.com/elastic/elasticsearch/pull/84165/commits/42d504a70a8ef52a87518dae665871679048f1d9

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java:85]: I think we should preserve order here too? Or perhaps even switch to an ordered map (`TreeeMap`) to ensure a consistent ordering?

I wonder if we even need to change the serialization here? I think we are sending the `externalId` twice now and might as well send it as a list and then reconstruct it on the receiver side?

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml
Comment (henningandersen) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml:195]: :+1: , but would be nice to even split this out into a separate PR? Or is the change too coupled with the rest of this PR?

**Review by henningandersen — COMMENTED** [review]
Did not get through it all, but wanted to relay a few early comments here.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml
Comment (fcofdez) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml:195]: Sure, let me split this PR 👍

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java:85]: I was thinking about that too. I would expect that the number of desired nodes would be fairly low in most cases. Let me revert this change.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml
Comment (fcofdez) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml:195]: I opened a new PR https://github.com/elastic/elasticsearch/pull/84227 with these changes. I think is better to hold this PR review until we get that one merged.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMetadata.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMetadata.java:79]: I think we should make this private now that we have the `Builder`?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMetadata.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMetadata.java:190]: nit: We often add a `builder()` method to the outer class. I like that, since it make the client code slightly shorter and reads slightly better I think.

**Review by henningandersen — COMMENTED** [review]
Sorry for the delay here. I wonder if we should make cluster membership a dynamically calculated property instead rather than persist it into metadata? For instance the current PR will IIUC keep all nodes as members if all nodes are abruptly shutdown and only some are started? Also at a higher level, I think knowledge about the current nodes in the cluster should not go into metadata to avoid persisting such information.

I can see a potential need for whether we ever saw a desired node, which would need to be in metadata. But I think the purpose here is different?

Left a couple random comments, not too important at this time.

**Comment by fcofdez** [comment]
@henningandersen I've followed a different approach where the membership is tracked in-memory, let me know if you prefer this approach. Thanks!

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipTracker.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipTracker.java:24]: I think this needs to be a service in order for various components to be able to ask it for the members? We can also turn it into a service when we add the first use of this.

**Review by henningandersen — COMMENTED** [review]
Thanks, this directions looks fine. I am not sure we need the quarantining/member removal though, I think we should simply remove that.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:57]: I think we should remove this. If we ever saw the node it is a member until a desired nodes update is made that removes it.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:35]: I wonder if we should just maintain this always, regardless of who is master? That would ensure it is readily available on failover. Otherwise, I worry that we need another special case for the failover case, i.e., when node was not previously master, since we only consider the delta of nodes below.

File: server/src/test/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipServiceTests.java
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipServiceTests.java:153]: I think we need a test demonstrating that a promoted master has the right set of members?

**Review by henningandersen — COMMENTED** [review]
A few more things around master failover and removing nodes.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:35]: Yes, I think it should be fine if we keep track of this in all nodes and as you mention it should simplify things

**Comment by fcofdez** [comment]
@elasticmachine run elasticsearch-ci/part-1 
(It was a Gradle connectivity issue)

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:39]: Do we need this condition? Seems more expensive than just clearing `members` always?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:79]:
```suggestion
    synchronized int trackedMembersCount() {
```

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by fcofdez** [comment]
@elasticmachine run elasticsearch-ci/part-2
(Unrelated LDAP failure)

**Comment by fcofdez** [comment]
Thanks Henning!

**Pull request merged**
